src/Symfony/Component/Cache/Adapter/MemcachedAdapter.php

```diff
@@ -11,68 +11,224 @@
+
 /**
  * @author Rob Frawley 2nd <rmf@src.run>
+ * @author Nicolas Grekas <p@tchwork.com>
  */
 class MemcachedAdapter extends AbstractAdapter
 {
+    private static $defaultClientOptions = array(
+        'persistent_id' => null,
+        'username' => null,
+        'password' => null,
+        'binary_protocol' => true,
+        'libketama_compatible' => true,
+    );
+
+    protected $maxIdLength = 250;
+
     private $client;
 
+    public static function isSupported()
+    {
+        return extension_loaded('memcached') && version_compare(phpversion('memcached'), '2.2.0', '>=');
+    }
+
     public function __construct(\Memcached $client, $namespace = '', $defaultLifetime = 0)
     {
+        if (!static::isSupported()) {
+            throw new CacheException('Memcached >= 2.2.0 is required');
+        }
+        $opt = $client->getOption(\Memcached::OPT_SERIALIZER);
+        if (\Memcached::SERIALIZER_PHP !== $opt && \Memcached::SERIALIZER_IGBINARY !== $opt) {
+            throw new CacheException('MemcachedAdapter: serializer must be PHP or IGBINARY.');
+        }
+        if (!$client->getOption(\Memcached::OPT_BINARY_PROTOCOL)) {
+            throw new CacheException('MemcachedAdapter: OPT_BINARY_PROTOCOL must be used.');
+        }
+        if ($client->getOption(\Memcached::OPT_NO_BLOCK)) {
+            throw new CacheException('MemcachedAdapter: OPT_NO_BLOCK must be disabled.');
+        }
+        $this->maxIdLength -= strlen($client->getOption(\Memcached::OPT_PREFIX_KEY));
+
         parent::__construct($namespace, $defaultLifetime);
         $this->client = $client;
     }
 
-    public static function isSupported()
+    /**
+     * @return \Memcached
+     *
+     * @throws \ErrorEception When invalid options or servers are provided.
+     */
+    public static function createConnection($servers, array $options = array())
     {
-        return extension_loaded('memcached') && version_compare(phpversion('memcached'), '2.2.0', '>=');
+        if (is_string($servers)) {
+            $servers = array($servers);
+        } elseif (!is_array($servers)) {
+            throw new InvalidArgumentException(sprintf('MemcachedAdapter::createClient() expects array or string as first argument, %s given.', get_type($servers)));
+        }
+        set_error_handler(function ($type, $msg, $file, $line) { throw new \ErrorException($msg, 0, $type, $file, $line); });
+        try {
+            if (!static::isSupported()) {
+                throw new trigger_error('Memcached >= 2.2.0 is required');
+            }
+            $options += static::$defaultClientOptions;
+            $client = new \Memcached($options['persistent_id']);
+            $username = $options['username'];
+            $password = $options['password'];
+            unset($options['persistent_id'], $options['username'], $options['password']);
+
+            // set client's options
+            foreach ($options as $name => $value) {
```

Comment:
I think reverse or not, it's the responsibility of the user to order correctly. I'd better behave the same as the native setOptions so that expectations don't have to be reversed when using our lib...